### PR TITLE
PP-832: PBSPro failover secondary server fails to continuously check …

### PR DESCRIPTION
**Pull request by prakashcv13**
…whether it needs to start a scheduler locally

#### Bug/feature Description
* [PP-832](https://pbspro.atlassian.net/browse/PP-832)

#### Affected Platform(s)
* All

#### Cause / Analysis / Design / Solution Description
* **[Issue and Solution description.](https://pbspro.atlassian.net/wiki/spaces/PD/pages/261193738/PP-832+PBSPro+failover+secondary+server+fails+to+continuously+check+whether+it+needs+to+start+a+scheduler+locally)**
* **[forum discussion](http://community.pbspro.org/t/pp-832-which-scheduler-to-talk-to-while-taking-over-from-primary/889)**

#### Testing logs/output
* *Attached to [PP-832](https://pbspro.atlassian.net/browse/PP-832)*

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [x] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/server/pbsd_main.c [hunk 1]
```diff
@@ -1450,6 +1450,11 @@ main(int argc, char **argv)
 		/* in case secondary didn't remove the file */
 		/* also tells the secondary to go idle	    */
 		(void)unlink(path_secondaryact);
+		/* 
+		 * Make the scheduler (re)-read the configuration
+		 * and fairshare usage.
+		 */
+		(void)contact_sched(SCH_CONFIGURE, NULL, pbs_scheduler_addr, pbs_scheduler_port);
```
Comment (suresh-thelkar) [inline src/server/pbsd_main.c]: Don't we need to check if contact_sched is indeed successful here? What if contact_sched fails and at this point of time if scheduler is down.

Comment (suresh-thelkar) [reply src/server/pbsd_main.c]: Also here we are sending SCH_CONFIGURE to re-read the configuration and fairshare usage. But EDD says we send SCH_SCHEDULE_FIRST. Does EDD needs an update or Am I missing anything ?

Comment (prakashcv13) [reply src/server/pbsd_main.c]: @suresh-thelkar - I have updated the EDD.
about checking if contact_sched being successful here, I feel there is no need.
Scheduler is started by the script or verified by the init script to be already running.
This part of the code is for primary, and we do not have failover for scheduler process.

Comment (suresh-thelkar) [reply src/server/pbsd_main.c]: @prakashcv13,
Thanks for updating the EDD. On not checking the return value of contact_sched I agree with you if we start primary server always with the init script. If this is the case I feel it is better to mention this assumption also in the EDD.  But I don't think this this is the case always for example consider the case of Windows and also another case of while the primary is up , stop it using qterm and after a while bring it up using the script pbs_server alone.

File: src/server/pbsd_main.c [hunk 2]
```diff
@@ -1907,27 +1907,25 @@ main(int argc, char **argv)
...
+				 * this is to get PBS_CONF_FILE if specified.*/
 				workenv = environ;
 				environ = origevp;
 #ifdef WIN32
-				strcpy(schedcmd, "net start pbs_sched");
-				sprintf(log_buffer, "starting scheduler: %s", schedcmd);
-				log_event(PBSEVENT_SYSTEM | PBSEVENT_FORCE,
-					PBS_EVENTCLASS_SERVER, LOG_CRIT,
-					msg_daemonname, log_buffer);
+				strncpy(schedcmd, sizeof(schedcmd), "net start pbs_sched");
```
Comment (bhroam) [inline src/server/pbsd_main.c]: Either use a snprintf() or put a '\0' in the last character of schedcmd.  I'd go with the snprintf() myself, but it is up to you.

**Comment by prakashcv13** [comment]
thank you @bhroam.

**Comment by prakashcv13** [comment]
@PBSPro/pbspro-maintainers - I have got necessary approvals, squashed my commits and rebased my branch... request you to take it forward.

**Comment by subhasisb** [comment]
@prakashcv13 it is still showing that it is out of date with base. Please check

**Comment by prakashcv13** [comment]
@subhasisb - thank you! yes, it seems I missed out on rebasing... did it now.

**Pull request merged**